msal/src/test/java/com/microsoft/identity/client/e2e/shadows/ShadowBaseController.kt

```diff
@@ -0,0 +1,74 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.client.e2e.shadows
+
+import com.microsoft.identity.common.java.cache.ICacheRecord
+import com.microsoft.identity.common.java.commands.parameters.SilentTokenCommandParameters
+import com.microsoft.identity.common.java.controllers.BaseController
+import com.microsoft.identity.common.java.providers.oauth2.OAuth2Strategy
+import com.microsoft.identity.common.java.providers.oauth2.OAuth2TokenCache
+import com.microsoft.identity.common.java.result.AcquireTokenResult
+import org.robolectric.annotation.Implementation
+import org.robolectric.annotation.Implements
+import org.robolectric.annotation.RealObject
+import org.robolectric.shadow.api.Shadow
+import org.robolectric.util.ReflectionHelpers.ClassParameter
+
+@Implements(BaseController::class)
+class ShadowBaseController {
+    companion object {
+        private var onRenewAccessTokenInvokedCallback: () -> Unit = {}
+
+        fun setOnRenewAccessTokenInvokedCallback(callback: () -> Unit) {
+            onRenewAccessTokenInvokedCallback = callback
+        }
+    }
+
+    @RealObject
+    private lateinit var baseController: BaseController
+
+    @Implementation
+    fun renewAccessToken(
+        parameters: SilentTokenCommandParameters,
+        acquireTokenSilentResult: AcquireTokenResult,
+        tokenCache: OAuth2TokenCache<*, *, *>,
+        strategy: OAuth2Strategy<*, *, *, *, *, *, *, *, *, *, *, *, *>,
+        cacheRecord: ICacheRecord
+    ) {
+        onRenewAccessTokenInvokedCallback.invoke()
```

Comment:
Suggestion: Could it be cleaner to instead have a boolean variable `renewAccessTokenCalled` instead of having to explicitly declare a callback?

Reply:
Failing the test from within `ShadowBaseController` isn't right: it's the responsibility of the test to assert and fail. 
I could replace the callback with a static boolean that's kept inside `ShadowBaseController`, where the test verifies the value and fails the test if the value is not as expected. `ShadowBaseController` becomes a bit more than just a shadow then; it's used to manage/track state as well. Not sure if that belongs there.
Wdyt? do you see another solution?

Reply:
To clarify - I don't think the test should be failed from the controller. Just that it should hold the boolean reference, sort of like a spy would. 

Although I think I misunderstood how this shadow works - is this essentially being called when any instance of any base controller calls this method, instead of referring to a single base controller instance? If so, then I think your approach makes more sense.

Reply:
Correct. We don't know which instance if BaseController is used, as this is an E2E test that can't see the internal controller etc. initialisation. So can't use a real spy - for that you need the instance. 
Robolectric's Shadow works through reflection. Which I consider a bit of a code smell, but the alternative was to expose the original `authResult` as part of the sign in result (see first commit). Which is even more of a code smell. 
I'm open to further suggestions though!

Reply:
Since we're hitting a real tenant now, can we compare the tokens between each call to see if they've been refreshed?

Reply:
That's what's being done in the tests. Is that what you mean?